### DISCUSSION TITLE
Read request body twice or only once on both OnRecordBeforeCreateRequest and OnRecordAfterCreateRequest?

### DISCUSSION THREAD

**Discussion by helenfufu** [pocketbase/pocketbase#1597] [Q&A] [answered]
Hello! What is the best way to read/use the same request body in two different hooks, i.e. `OnRecordBeforeCreateRequest` and `OnRecordAfterCreateRequest` both intercepting a create request? (This question could also apply to using both `Before` and `After` hooks for update and delete, too.)

At the moment, I'd like to do the following:
1. In `OnRecordBeforeCreateRequest`, intercept the create request body, query a external API, then set some additional record fields before the record gets created. I have this part working, with something like the code block at the bottom of this post.
2. In `OnRecordAfterCreateRequest`, use the same create request body to update the `users` collection's `custom_profiles` relation with the newly created custom profile under the incoming request's `user_id` (which corresponds to `users.id`). While writing this, I've realized that reading the `httpContext.Request().Body`, either directly as shown below or with `httpContext.Bind` (which appears to do the same thing under the hood), gives 400 `EOF` errors, because we already read from the same `io.Reader` previously.

I am not super familiar with `echo` so not sure if there's an easy way to do this using the framework. Typically, I would just store the unmarshalled result in a variable to use later on in a single handler function, or pass it through to functions called down the line, but I'm not quite sure how to do it with the two separate hooks each taking in `e *core.RecordCreateEvent` as input. I suppose we can use this `io.TeeReader` suggestion [here](https://stackoverflow.com/questions/39791021/how-to-read-multiple-times-from-same-io-reader), but it still seems a bit hacky to read the same request twice. Is there a better way to access the same request body in `Before` and `After` hooks on the same request (i.e. only reading it once)?

Reference code:
```go
func main() {
	app := pocketbase.New()

	app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
		collection := e.HttpContext.PathParam(apis.ContextCollectionKey)

		switch collection {
		case "custom_profiles":
			return populateCustomProfiles(app, e.HttpContext, e.Record)
		}
		return nil
	})

	app.OnRecordAfterCreateRequest().Add(func(e *core.RecordCreateEvent) error {
		collection := e.HttpContext.PathParam(apis.ContextCollectionKey)

		switch collection {
		case "custom_profiles":
			return updateUserCustomProfiles(app, e.HttpContext, e.Record)
		}
		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}

type createCustomProfileRequest struct {
	UserId string `json:"user_id"`
	Name string `json:"name"`
}


func populateCustomProfile(app core.App, httpContext echo.Context, record *models.Record) error {
	request := httpContext.Request()
	bodyBytes, err := ioutil.ReadAll(request.Body)
	if err != nil {
		return err
	}
	var body createCustomProfileRequest
	if err := json.Unmarshal(bodyBytes, &body); err != nil {
		return err
	}
	name := strings.ToLower(body.Name)

	customProfileFields, err := getAdditionalCustomProfileFields(name)
	if err != nil {
		return err
	}
	record.Set("field1", customProfileFields.Field1)
	record.Set("field1", customProfileFields.Field2)
	
	return nil
}

func updateUserCustomProfiles(app core.App, httpContext echo.Context, record *models.Record) error {
	request := httpContext.Request()
	bodyBytes, err := ioutil.ReadAll(request.Body)
	if err != nil {
		return err
	}
	var body createCustomProfileRequest
	if err := json.Unmarshal(bodyBytes, &body); err != nil {
		return err
	}
	// TODO: Add update logic once we have the request body.
	return nil
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
You can get a parsed copy of the json request using the `apis.RequestData(c echo.Context)` helper. The body/data fieds will be available under the `Data` key.

**Reply by helenfufu** [reply]
Ah, thanks for the advice (and all your work on PocketBase)! Got it working 😄
